docs/branching-model.md

```diff
@@ -0,0 +1,57 @@
+# Git branching model
+
+## Master branch
+
+The default branch for a project is the master branch. This branch must always be in a stable state: the build must be successful, all tests must pass, style must follow the [style guide](cplusplus-style-guide.md) and static analysis should not give warnings or errors.
+
+Commits representing a release will be tagged as described in the [release strategy](release-strategy.md#tagging-and-version-numbers).
+
+
+## Development forks
+
+In order to keep the main repo clear from development branches, developers must create a fork of any project they are planning on contributing to. In addition, for projects which are or will be open sourced this is the easiest way to begin developing.
+The only scenario when it may be currently necessary to push branches directly to the repository is where the branch changes the Jenkinsfile. These should not be picked up from a fork for security reasons.
+
+Once the fork has been created, it is expected that the user would create a branch to track each new feature, bugfix, etc,  being made. If any of these tasks are complex, they can be addressed in multiple increments that are individually reviewed and merged. See [Integration (pull requests)](#integration-pull-requests) for more information about merging development branches.
+
+A task is only considered done once all the branches associated with it have been merged into the master branch; this encourages tasks to be well defined so they can be assessed and merged quickly, with the minimum amount of code to be integrated every time.
+
+Long-lived development branches are strongly discouraged but if necessary, its creation must be discussed to try to find alternatives. Long-lived branches tend to strongly diverge from the master branch, making the review and integration complicated and tedious.
+
+To further aid clarity we advise using the following prefixes based on the nature of the change being made:
+
+ * `feature`: New feature
+ * `bugfix`: Fixing a bug in the code
+ * `refactor`: Not a change in functionality, but a general improvement to code health
+ * `backport`: Porting a fix from the [master branch](#master-branch) to an existing [release branch](#release-branches)
+
+Jira ticket numbers must be present in the branch names in order for the Jira <-> GitHub plugin to link them automatically. But they must not be used as the whole branch name, because they are not descriptive enough out of that context.
+
+**Examples**
+
+ * `feature/ldgr-111-transaction-fees` - A branch for the implementation of transaction-based fees
+ * `bugfix/ldgr-222-http-empty-buffer` - A branch for fixing a specific issue
+ * `refactor/ldgr-333-update-state-machine` - A branch for updating a state machine
```

Comment:
Note that Jira has a limitation on the branch names it will handle of 50 characters. This naming convention now consumes around 20 of them.

Reply:
That's a good point, this suggestion was based on what some people were already doing and it worked nicely. Another option would be to label the PRs as I think your team is now doing. That would shorten the names and display a pretty distinctive icon in Jira.

Since you were the ones that suggested the branch name, would you be happy with this change @bagnall-fetch @mattmcd ?

Reply:
`feature` and `bugfix` is more familiar to me although I take the point about JIRA name length limitation.  I don't think `dev` is descriptive enough though.

Reply:
Sorry, my question was about replacing the `ldgr-111` in the branch name with a PR title like `LDGR-111 Implementation of transaction-based fees`.

Reply:
Oh, got it.  So instead of (a) `feature/ldgr-111-transaction-fees` have (b) `feature/LDGR-111 Implementation of transaction-based fees`?
I favour (a) for branch names, and keep under 50 chars as Katie suggests, because it's shorter so displays better in IDE,JIRA,etc and ticket number can link back to more detailed description in JIRA.  Also spaces in branch names don't feel right to me for some reason.

Reply:
We ended up discussing this offline and we were both in agreement :) The idea is to have the following:
* Branch name: feature/transaction-fees
* PR title: LDGR-111 Implementation of transaction-based fees

Reply:
How about:
* PR title MUST have ticket number in it (to link back to JIRA)
* branch SHOULD have ticket number in it - looks like some people favour longer branch names vs others (e.g. me) who like to be able to see all the commits relating to a ticket in JIRA.